### DISCUSSION TITLE
Make easier way to manage version changes - npm refactor?

### DISCUSSION THREAD

**Discussion by ninjasort** [npm/feedback#820] [General]
Is there any way npm could build something that **makes version upgrades easier**?

It would be great if npm had a tool to automatically reference a changelog and automatically find references to outdated code based on annotations or decorators. This way we could run something like `npm refactor` and get a summary of changelog-style tests that would allow us to write assertions based on changes to a package version.

This way we know that we are getting all checks and balances in place when a new release comes out without needing to sift through the random documentation across packages. Basically introduce a way to standardize it.

**Comment by ljharb** [comment]
This doesn’t even seem like something that’s possible to do reliably in JavaScript, let alone something npm should be involved in (npm packages can be many languages and formats, not just JavaScript).

**Comment by ninjasort** [comment]
Here's more of an example of what I mean...

This is a gatsby package that had a breaking change, and they provided docs for how to upgrade your own package.

![image](https://user-images.githubusercontent.com/1614348/210183397-3c60481f-4d87-4b86-8a04-8e971b36d63a.png)

Imagine a package decides to upgrade to a new major version with many breaking changes.

It would be useful if npm could provide a better developer experience for adopting those breaking changes. 

At the very minimum it should provide a step by step checklist to confirm the labeled changes (provided by dependency author) were adopted and bump your package version automatically.

For example, with [semantic-release](https://github.com/semantic-release/semantic-release#how-does-it-work) you can use a commit message to update the version of your package. 

If npm had `npm refactor`, it could provide a prompt similar to [commitizen](https://github.com/commitizen/cz-cli#if-your-repo-is-commitizen-friendly) to refactor a package based on dependency upgrades.

![image](https://user-images.githubusercontent.com/1614348/210183655-6823ce6e-8a3d-441d-837d-865fc1cbea9f.png)

Ideally, `npm upgrade` would actually behave like [npm-check-updates](https://github.com/raineorshine/npm-check-updates#usage) and upgrade the dependency version numbers.

![image](https://user-images.githubusercontent.com/1614348/210183668-b383624e-ca1e-4301-ac55-837c8fdfa63a.png)


Then you can run an `npm refactor` to step through the code changes and auto commit your own package upgrade.

Makes sense?

**Reply by ljharb** [reply]
That's a lot of conventions for npm to attempt to impose on an ecosystem that hasn't come anywhere close to agreeing on what conventions to follow - even semantic-release and its associated conventions aren't anywhere close to a majority of usage.
